python/taichi/lang/matrix.py

```diff
@@ -76,57 +76,29 @@ def __init__(self,
                 self.m = 1
             self.entries = [x for row in mat for x in row]
 
-        # construct global matrix
         else:
-            self.entries = []
-            self.n = n
-            self.m = m
-            self.dt = dt
             if dt is None:
-                for i in range(n * m):
-                    self.entries.append(impl.expr_init(None))
-            else:
-                assert not impl.inside_kernel()
-                for i in range(n * m):
-                    self.entries.append(impl.field(dt))
-                self.grad = self.make_grad()
-
-            if layout is not None:
-                assert shape is not None, 'layout is useless without shape'
-            if shape is not None:
-                if isinstance(shape, numbers.Number):
-                    shape = (shape, )
-                if isinstance(offset, numbers.Number):
-                    offset = (offset, )
-
-                if offset is not None:
-                    assert len(shape) == len(
-                        offset
-                    ), f'The dimensionality of shape and offset must be the same  ({len(shape)} != {len(offset)})'
-
-                import taichi as ti
-                if layout is None:
-                    layout = ti.AOS
-
-                dim = len(shape)
-                if layout.soa:
-                    for i, e in enumerate(self.entries):
-                        ti.root.dense(ti.index_nd(dim),
-                                      shape).place(e, offset=offset)
-                        if needs_grad:
-                            ti.root.dense(ti.index_nd(dim),
-                                          shape).place(e.grad, offset=offset)
-                else:
-                    var_list = []
-                    for i, e in enumerate(self.entries):
-                        var_list.append(e)
-                    if needs_grad:
-                        for i, e in enumerate(self.entries):
-                            var_list.append(e.grad)
-                    ti.root.dense(ti.index_nd(dim),
-                                  shape).place(*tuple(var_list), offset=offset)
+                # create a local matrix with specific (n, m)
+                self.entries = [impl.expr_init(None) for i in range(n * m)]
+                self.n = n
+                self.m = m
```

Comment:
1. Is `dt is None` case useful?
2a. If 1 is yes, note that the original logic still handles the layout/shape in this case (i.e. what you've put into `Matrix.field`). Yet the new logic won't do that anymore. Is this correct?
2b. Otherwise, shall we just enforce `dt` to never be `None`?

Reply:
1. IMO `dt is None` is only used for local matrices.
2a. I think **the old logic is incorrect** at the first place @yuanming-hu, `ti.Matrix(n, m, dt=None, shape=233)`? How to allocate a matrix with `None` type ahead?
2b. SGTM, so we should use no default value `None` for dtype, see below.

Reply:
I see. So `ti.Matrix(3,4)` is a valid use case, i guess?